alf/utils/dist_utils.py

```diff
@@ -361,23 +359,84 @@ def estimated_entropy(dist,
+            assert torch.all(torch.isfinite(entropy))
+        entropy = entropy.mean(dim=0)
         entropy_for_gradient = entropy
     else:
-        entropy = -dist.log_prob(tf.stop_gradient(single_action))
+        entropy = -dist.log_prob(single_action.detach())
         if check_numerics:
-            entropy = tf.debugging.check_numerics(entropy, 'entropy')
-        entropy_for_gradient = -0.5 * tf.math.square(entropy)
-        entropy = tf.reduce_mean(entropy, axis=0)
-        entropy_for_gradient = tf.reduce_mean(entropy_for_gradient, axis=0)
+            assert torch.all(torch.isfinite(entropy))
+        entropy_for_gradient = -0.5 * entropy**2
+        entropy = entropy.mean(dim=0)
+        entropy_for_gradient = entropy_for_gradient.mean(dim=0)
     return entropy, entropy_for_gradient
 
 
+def entropy_with_fallback(distributions, action_spec):
+    """Computes total entropy of nested distribution.
+    If entropy() of a distribution is not implemented, this function will
+    fallback to use sampling to calculate the entropy. It returns two values:
+    (entropy, entropy_for_gradient).
+    There are two situations:
+    * entropy() is implemented. entropy is same as entropy_for_gradient.
+    * entropy() is not implemented. We use sampling to calculate entropy. The
+        unbiased estimator for entropy is -log(p(x)). However, the gradient of
+        -log(p(x)) is not an unbiased estimator of the gradient of entropy. So
+        we also calculate a value whose gradient is an unbiased estimator of
+        the gradient of entropy. See estimated_entropy() for detail.
+    Example:
+        ent, ent_for_grad = entropy_with_fall_back(dist, action_spec)
+        alf.summary.scalar("entropy", ent)
+        ent_for_grad.backward()
+    Args:
+        distributions (nested Distribution): A possibly batched tuple of
+            distributions.
+        action_spec (nested BoundedTensorSpec): A nested tuple representing the
+            action spec.
+    Returns:
+        tuple of (entropy, entropy_for_gradient). You should use entropy in
+        situations where its value is needed, and entropy_for_gradient where
+        you need to calculate the gradient of entropy.
+    """
+
+    def _calc_outer_rank(dist: td.Distribution, action_spec):
+        if isinstance(dist, td.TransformedDistribution):
+            # TransformedDistribution does not implement the two necessary
+            # interface functions of Distribution. So we have to use the
+            # original distribution it transforms.
+            # TransformedDistribution is used by NormalProjectionNetwork with
+            # scale_distribution=True
+            dist = dist.base_dist
+        return (len(dist.batch_shape) + len(dist.event_shape) - len(
+            action_spec.shape))
+
+    def _compute_entropy(dist: td.Distribution, action_spec):
+        if isinstance(dist, td.TransformedDistribution):
+            entropy, entropy_for_gradient = estimated_entropy(dist)
```

Comment:
It might be possible to get a closed-form of entropy given a Normal as the base dist? It's better (lower variance) than this estimated one. Also, it seems that after transformations, the entropy actually changes, and we are computing the entropy of the base dist instead of the input dist.

Reply:
Something like what TFP does: https://github.com/tensorflow/probability/blob/356cfddef026b3339b8f2a81e600acd2ff8e22b4/tensorflow_probability/python/distributions/transformed_distribution.py#L636
(Probably it's complicated, but we need to spend time figuring out if the current estimation is the best way to do this).

Reply:
Noted.

Reply:
the entropy for transformed distribution is only available for affine transformation as noted in the linked code. Because of this, we can make the condition as the following:
```python
if (isinstance(dist, TransformedDistribution) and
   not (dist.bijector.is_constant_jacobian
           and  dist.bijector._is_injective)):
```
And we may remove `SqushSpecNormal`

Reply:
You can take a look at the code of TransformedDistribution. It can only calculate entropy if (dist.bijector.is_constant_jacobian and  dist.bijector._is_injective)

Reply:
Replying thru chat.